### ISSUE TITLE
thread 'vcpu-1' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidCompQueue(0)'

### ISSUE THREAD

**Issue by jmpesp** [oxidecomputer/propolis#308]
saw the following:

```
thread 'vcpu-1' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidCompQueue(0)', lib/propolis/src/hw/nvme/mod.rs:306:44
stack backtrace:
   0:          0x1d01c4c - std::backtrace_rs::backtrace::libunwind::trace::h9b534fc6095f520f
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:          0x1d01c4c - std::backtrace_rs::backtrace::trace_unsynchronized::hfcd8a890f991a10d
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:          0x1d01c4c - std::sys_common::backtrace::_print_fmt::hdf432e5f0940fa9c
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys_common/backtrace.rs:65:5
   3:          0x1d01c4c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h5693a803500ade80
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys_common/backtrace.rs:44:22
   4:          0x1d5937a - core::fmt::write::h94f582bdfeeeb1a6
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/fmt/mod.rs:1209:17
   5:          0x1cf2a34 - std::io::Write::write_fmt::hd3f00507505f204d
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/io/mod.rs:1682:15
   6:          0x1d01a20 - std::sys_common::backtrace::_print::h21d778336866aef0
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys_common/backtrace.rs:47:5
   7:          0x1d01a20 - std::sys_common::backtrace::print::h8238ed32ff302292
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys_common/backtrace.rs:34:9
   8:          0x1d04786 - std::panicking::default_hook::{{closure}}::h933e4cf16faea6e2
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:267:22
   9:          0x1d0441c - std::panicking::default_hook::h9576f50ee054cc3c
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:286:9
  10:          0x1d050ac - std::panicking::rust_panic_with_hook::hb09aca5ad1ad2161
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:688:13
  11:          0x1d04e25 - std::panicking::begin_panic_handler::{{closure}}::h189c984e260f383e
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:579:13
  12:          0x1d020f0 - std::sys_common::backtrace::__rust_end_short_backtrace::hcab55adb02d49332
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys_common/backtrace.rs:137:18
  13:          0x1d04b51 - rust_begin_unwind
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:575:5
  14:          0x1d55d03 - core::panicking::panic_fmt::hc653fbe903263e77
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/panicking.rs:65:14
  15:          0x1d56215 - core::result::unwrap_failed::hf3fb563cfc573601
                               at /rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/result.rs:1791:5
  16:          0x13185bc - propolis::hw::nvme::PciNvme::reg_ctrl_write::h14ab8780afeb89e3
  17:          0x13dd205 - propolis::util::regmap::RegMap<ID>::process::h52e6f510ee8fdd59
  18:          0x1318fd7 - <propolis::hw::nvme::PciNvme as propolis::hw::pci::device::Device>::bar_rw::hc5d11f72e3214f6f
  19:           0xf2794c - propolis::hw::pci::device::<impl propolis::hw::pci::Endpoint for D>::bar_rw::he4d8858630c4f422
[ Feb  7 07:18:42 Stopping because all processes in service exited. ]
[ Feb  7 07:18:42 Executing stop method (:kill). ]
```

**Comment by jmpesp** [comment]
This was running an 8 vCPU instance with 16G of RAM, with one attached nvme disk backed by Crucible

**Comment by luqmana** [comment]
So we tried to unwrap some result and got an `InvalidCompQueue(0)` error. That comes from a call to one of `NvmeCtrl::{create,delete,get}_cq`. `0` at least tells us it was the Admin CQ (`ADMIN_QUEUE_ID == 0`). Don't see it in the backtrace (I'm assuming due to inlining) but my guess is a call to `NvmeCtrl::get_admin_cq` tripped this:
```rust
/// Returns a reference to the Admin [`CompQueue`].
///
/// # Panics
///
/// Panics if the Admin Completion Queue hasn't been created yet.
fn get_admin_cq(&self) -> Arc<CompQueue> {
    self.get_cq(queue::ADMIN_QUEUE_ID).unwrap()
}
```

It looks like the guest tried to write to one of the admin queue doorbell registers (either `DoorBellAdminCQ` or `DoorBellAdminSQ`) at a point where we hadn't yet set them up?

@jmpesp Do you have the rest of the log available by any chance? It'd be useful to know if this was during boot/shutdown or something. Could be some race in shutting down the controller and servicing outstanding requests.

In theory the guest shouldn't be trying to hit those register if the controller (and therefore the admin queues) hasn't been initialized. While useful during development, we probably shouldn't crash in such cases and try to be resilient. I can put up a patch to not panic but it'd be good to know if this is reproducible and indicative of a bigger issue.

**Comment by rmustacc** [comment]
> In theory the guest shouldn't be trying to hit those register if the controller (and therefore the admin queues) hasn't been initialized. While useful during development, we probably shouldn't crash in such cases and try to be resilient. I can put up a patch to not panic but it'd be good to know if this is reproducible and indicative of a bigger issue.

I do think it's pretty important that we assume every register in every BAR can be read or written to from the guest no matter the state and that should not crash us. So not crashing here will be pretty important.

**Comment by jmpesp** [comment]
> @jmpesp Do you have the rest of the log available by any chance? It'd be useful to know if this was during boot/shutdown or something. Could be some race in shutting down the controller and servicing outstanding requests.

This wasn't during boot or shutdown, no. I didn't attach the log because it was only filled with those MSR related INFO messages :)

**Comment by luqmana** [comment]
I'll close this for now as the panic has been addressed by #309.
